parquet-thrift/src/main/java/org/apache/parquet/thrift/ThriftRecordConverter.java

```diff
@@ -393,20 +533,42 @@ public long readI64() throws TException {
         }
       });
     }
-
   }
 
-  /**
-   * converts Binary into String
-   * @author Julien Le Dem
-   *
-   */
-  static class FieldStringConverter extends PrimitiveConverter {
+  static final class StringConverter extends PrimitiveConverterBase {
 
-    private final List<TProtocol> events;
+    StringConverter(List<TProtocol> events) {
+      super(events);
+    }
 
-    public FieldStringConverter(List<TProtocol> events, ThriftField field) {
-      this.events = events;
+    @Override
+    protected ParquetProtocol decodeFromDictionary(Dictionary d, int i) {
+      final Binary decoded = d.decodeToBinary(i);
+
+      return new ParquetProtocol("readString() binary") {
+        // since we don't know whether this field is a binary or a string,
+        // we lazily convert the first time readString() or readBinary is called.
+        // we can't eagerly convert because a binary blob might not be a valid string
+        // TODO: ensure that this code doesn't need to be threadsafe
```

Comment:
I think you're right.
Parquet in hadoop doesn't need to be thread safe, but idk about when its used in spark and similar things.